### PR TITLE
Ports - Kasey and Elle

### PR DESCRIPTION
**Pull request by dev-elle-up**
# slack.rb
Congratulations! You're submitting your assignment!

You and your partner should _collaborate_ on the answers to these questions.

## Comprehension Questions
Question | Answer
---      | ---
How did you go about exploring the Slack API? Did you learn anything that would be useful for your next project involving an API? | We worked individually to explore the Slack API primarily by reading the documentation. We also used Postman to become familiar with the structure of sending GET and POST requests and examine the JSON that was returned.
Give a short summary of the _request/response cycle_. Where does your program fit into that scheme? | In the request/response cycle, a request is sent to the server that includes query parameters. The server returns a response that contains a header and a body. Our program follows this scheme by first sending a query to the Slack API as a GET request. When the response is received, our program stores the data returned in a variable and then parses the data for use in our program. Similarly, we use a POST request to send a message (request). The response returned contains a parameter which indicates whether the action was successful; we use this in our program to raise an error if it was not successful.
How does your program check for and handle errors when using the Slack API? | We check the response returned by the post request for the parameter "ok" and raise an error if this value is false. These errors are mostly useful for being detected by our tests.
Did you need to make any changes to the design work we did in class? If so, what were they? | We decided not to have Channel and User inherit from a Recipient class. 
Did you use any of the _inheritance idioms_ we've talked about in class? How? | We used the pattern of having the SlackError inherit from StandardError.
How does VCR aid in testing a program that uses an API? | VCR reduces the number of API calls your program makes to the server by "recording" the results in a .yml file and then spoofing an API call by playing back the results of the last (recorded) call. This is helpful because API calls may be limited or charged for. It also helps by letting you continue to work if your computer or the server is offline.

### CODE REVIEW

File: lib/slack.rb [hunk 1]
```diff
@@ -1,11 +1,158 @@
...
+
+require_relative "channel"
+require_relative "user"
+require "httparty"
+require "dotenv"
+Dotenv.load
+
+class SlackError < StandardError; end
+
+class Slack
```
Comment (CheezItMan) [inline lib/slack.rb:12]: This class would be better named `SlackWorkspace` since slack can include multiple workspaces.  

Also it would be good to wrap everything in a module

File: lib/slack.rb [hunk 2]
```diff
@@ -1,11 +1,158 @@
...
+  attr_reader :channels, :users
+
+  def initialize()
+    @channels = []
+    @users = []
+    get_lists
+  end
+
+  def get_lists()
```
Comment (CheezItMan) [inline lib/slack.rb:21]: Just notice here how the loops look almost identical.  This is a good indication that you can dry things up using duck typing / polymorphism.

File: lib/slack.rb [hunk 3]
```diff
@@ -1,11 +1,158 @@
...
+  end
+
+  def self.send_msg(recipient, text)
+    if recipient == nil
+      puts "Please select a channel or user first."
+      main
+    end
+
+    url = "https://slack.com/api/chat.postMessage"
+    if recipient.class == Channel
```
Comment (CheezItMan) [inline lib/slack.rb:69]: The `if` and `elsif` here are identical, so you just need to check to see if `recipient` is not `nil`.

File: lib/slack.rb [hunk 4]
```diff
@@ -1,11 +1,158 @@
...
+      return true
+    else
+      raise SlackError, "Error when posting '#{text}' to #{recipient}, error: #{response["error"]}"
+    end
+  end
+end
 
 def main
-  puts "Welcome to the Ada Slack CLI!"
+  slack = Slack.new
```
Comment (CheezItMan) [inline lib/slack.rb:96]: It would be better to have `main` and the `Slack` class in separate files.

File: specs/slack_spec.rb [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        expect(return_value).must_be_kind_of Slack
+        expect(return_value.channels[0].slack_id).must_equal "CH41RMLP4"
+        expect(return_value.users[1].user_name).must_equal "cello.elle"
+      end
+    end
+  end
+
+  describe "select users and channels" do
+    it "select a user" do
+      VCR.use_cassette("slack_query") do
```
Comment (CheezItMan) [inline specs/slack_spec.rb:21]: You should also test with an invalid user.

File: specs/slack_spec.rb [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        slack = Slack.new # <-- .yml filename
+        user_a = slack.select_user("UH53ZCBBR") # <-- this is "id" value for "kaseea"
+        user_b = slack.select_user("kaseea") # <-- this is "name" value for "kaseea"
+        # binding.pry
+        expect(user_a.user_id).must_equal "UH53ZCBBR"
+        expect(user_b.user_name).must_equal "kaseea"
+      end
+    end
+
+    it "select a channel" do
```
Comment (CheezItMan) [inline specs/slack_spec.rb:31]: ditto here

File: specs/slack_spec.rb [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        slack = Slack.new
+        channel_a = slack.select_channel("apiiiii")
+        channel_b = slack.select_channel("CH4BCTLHK") # <-- this is the "id" value for the general channel
+        expect(channel_a.name).must_equal "apiiiii"
+        expect(channel_b.name).must_equal "random"
+      end
+    end
+  end
+
+  describe "select users and channels" do
```
Comment (CheezItMan) [inline specs/slack_spec.rb:42]: This describe seems to include seeing details rather than select user and channels.

You should also test what happens when no user or channel is selected.  

You should test with both a user and channel.

File: specs/slack_spec.rb [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+      VCR.use_cassette("slack_query") do
+        slack = Slack.new
+        channel_a = slack.select_channel("apiiiii")
+        user_b = slack.select_user("kaseea")
+        expect(channel_a).must_respond_to :details
+        expect(user_b).must_respond_to :details
+      end
+    end
+
+    describe "send_msg" do
```
Comment (CheezItMan) [inline specs/slack_spec.rb:53]: Good edge-case tests here.

File: specs/user_spec.rb [hunk 9]
```diff
@@ -0,0 +1,15 @@
+require_relative "test_helper"
```
Comment (CheezItMan) [inline specs/user_spec.rb:1]: This file doesn't seem to fit here.  Also where are tests on the `User` class.

**Comment by CheezItMan** [comment]
# slack.rb
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene (no slack tokens) | Good number of commits and good commit messages except one detailing a wave.
Comprehension questions | Check, nicely detailed answer on the request-response cycle
**Functionality** |
List users/channels | Check
Select user/channel | Check
Show details | Check
Send message | Check
Program runs without crashing | Check, well done
**Implementation** |
API errors are handled appropriately | Check
Inheritance model matches in-class activity | No, y'all made different choices
Inheritance idioms (abstract class, template methods, polymorphism) used appropriately | Not so much.
Methods are used to break work down into simpler tasks | Check
Class and instance methods are used appropriately | Check
**Overall** |  Well done, you hit most of the learning goals.  Your design is pretty good, but you missed some opportunities to take advantage of inheritance and duck typing to simplify some of your code.  See the instructor reference design and take a look at my inline comments.  Let me know if you have any questions.

File: lib/slack.rb [hunk 10]
```diff
@@ -1,11 +1,158 @@
...
+  end
+
+  def get_lists()
+    query_parameters = {
+      token: ENV["SLACK_API_TOKEN"],
+      pretty: 1,
+    }
+    url = "https://slack.com/api/conversations.list?"
+    response = HTTParty.get(url, query: query_parameters)
+    response["channels"].each do |x|
```
Comment (CheezItMan) [inline lib/slack.rb:28]: .map could also be used here to give you an array of Channel instances
